core/modules/filter/tests/filter_dialog.test

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * @file
+ * Functional tests for the Filter module.
+ */
+
+/**
+ * Test for filter dialog link validation.
+ */
+class FilterEditorLinkValidateTestCase extends BackdropWebTestCase {
+
+  protected $profile = 'testing';
+  protected $editorUser;
+  protected $format;
+
+  /**
+   * Set up testing environment.
+   */
+  public function setUp() {
+    parent::setUp(
+      'filter',
+      'ckeditor',
+      'filter_formtest',
+      'language',
+      'locale',
+    );
+
+    // Add a language and set it as default.
+    $german = (object) array(
+      'langcode' => 'de',
+      'name' => 'German',
+      'direction' => LANGUAGE_LTR,
+    );
+    language_save($german);
+    config_set('system.core', 'language_default', 'de');
+
+    // Create text format using filter default settings.
+    $filter_info = filter_filter_info();
+    $filters = array_fill_keys(array_keys($filter_info), array());
+    $format = (object) array(
+      'format' => 'filter_custom',
+      'name' => 'Filter custom',
+      'filters' => $filters,
+      'editor' => 'ckeditor',
+    );
+    filter_format_save($format);
+    $permission = filter_permission_name($format);
+    $this->format = $format;
+
+    // Create content type.
+    $this->backdropCreateContentType(array('type' => 'page', 'name' => 'Page'));
+
+    // Create user and log in.
+    $this->editorUser = $this->backdropCreateUser(array(
+      'access content',
+      'bypass node access',
+      $permission,
+    ));
+    $this->backdropLogin($this->editorUser);
+  }
+
+  /**
+   * Check function _filter_format_editor_link_url_validate().
+   */
+  public function testEditorLinkValidate() {
+    // Create a node and get the nid for the calling path.
+    $node = $this->backdropCreateNode();
+    $calling_path = "node/{$node->nid}/edit";
+    $link_token = filter_editor_dialog_token($this->format, 'link', $this->editorUser, $calling_path);
+    $german = language_load('de');
+    $options = array(
+      'query' => array(
+        'token' => $link_token,
+        'calling_path' => $calling_path,
+      ),
+      'language' => $german,
+    );
+    $this->backdropGet('editor/dialog/link/' . $this->format->format, $options);
+
+    $edit = array(
+      'attributes[text]' => 'The link',
+    );
+    $invalid_urls = array(
+      'javascript://boo',
+      'http://foo>bar.baz',
+      'http://foobar\.baz',
+    );
+    foreach ($invalid_urls as $url) {
+      $edit['attributes[href]'] = $url;
+      $this->backdropPost(NULL, $edit, t('Save'));
+      $this->assertRaw('The URL <em class="placeholder">' . check_plain($url) . '</em> is not valid.', format_string('Invalid URL %url did not pass validation', array(
+        '%url' => $url,
+      )));
+    }
+
+    $valid_urls = array(
+      // Should get altered.
+      'node/1' => '/de/node/1',
+      'ärger-im-büro' => '/de/%C3%A4rger-im-b%C3%BCro',
+      '<front>?page=1#main-content' => '/de?page=1#main-content',
+      '#ärger' => '#%C3%A4rger',
```

Comment:
Hm, I understand the intention, but... `id="ärger"` wouldn't be valid, as far as I know.

Reply:
I'm not sure what you mean here. I think I tested it both ways.

Reply:
What I mean: it's not recommended that an id attribute contains non-ascii characters. I know, it's valid in html5, but browsers might not be as permissive. For sure there will be trouble with Javascript. So: should we really support that in the filter link dialog? Your decision. :wink:

Reply:
Well, it's not really my decision alone. I figured encoding the anchor would produce something valid. In my testing it does work well enough. However I haven't looked into how it works with JS. Do you have any more info on that?

Reply:
I don't have more info on that, it's just a suspicion. In html5/css3 all utf-8 characters seem to valid in class/id attributes, however, function [backdrop_clean_css_identifier](https://docs.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_clean_css_identifier/1) is pretty strict, as it still follows css2.1 standards. Regarding js - might need some testing, if jquery chokes on utf8 characters in selectors, but that might be out of scope of this issue/PR.

Reply:
Would you be able to elaborate how `backdrop_clean_css_identifier` comes into play here?

Here's what I tested manually with this PR:

1. I create a page and add a link with the dialog. When I type in `#ärger` it gets rewritten to `#%C3%A4rger`.
2. I add a lot of space and put `<p id="#%C3%A4rger">Anchor 1</p>`.
3. I create another link with the dialog. When I type in `#ärger2` it gets rewritten to `#%C3%A4rger2`.
4. I add a lot of space and put `<p id="#ärger2">Anchor 2</p>`.

Both anchors seem to work for me.

Whether the ID can be UTF-8 or not seems to be out of the scope of this issue. But by rewriting the hash URL it seems to at least make that part valid.

Perhaps I'm missing something?

Reply:
@herbdool it's still your decision. :wink:

The problem is something completely different.

Add a link with "#ärger", save the node. Open the node form again, edit that link - it's now "#%25C3%25A4rger". Save the node again, open the form, edit the link - it's now "#%2525C3%2525A4rger"... and so on.

The base problem is that the link editor widget re-encodes every time one edits a link. Probably that's not what you want.
But it will be hard, if not impossible to fix that.

Reply:
BTW: I'm not sure, if my PR fixed the re-re-encoding problem, but if I remember correctly, it should have partly. - Could be totally wrong, though.